### PR TITLE
Batches explore

### PR DESCRIPTION
**Pull request by 0xaaiden**
# Summary

Added recent batches tab to the Homepage table

Screenshot:
image

# To Test

* Open homepage, click on Recent Batches Tab.
* Check it display recent batch orders
* Click on Tx. Hash should redirect to order page
* Hovering over tokens should take to etherscan token page
* Embedded Menu should have 3 items (Etherscan, Cow Explorer, and copy Tx Id)

### CODE REVIEW

**Comment by fleupold** [comment]
This looks awesome! 

What is the logic for calculating the batch value? I noticed for a lot of them it says 0 even if e.g. ETH is involved. Can we infer the batch value if we have the price for at least one side of each trade? Also, in case we can't can we use something like "prices unknown" or similar rather than $0 ?

**Comment by elena-zh** [comment]
Hey @0xaaiden , great job!
 I have done some tests and here are my notes: 
1. It would be great to start the 2nd word from a lower-case letter (as everywhere in the app)
low case
small

2. It would be great to crop a TX hash in the middle 
cropp in the middle
3. It would be great to show a public name tag for a token (and give a bit more context for a token instead of showing 1 letter) 
fetch token name
tracker
4. It would be great to show a downward arrow near the Date column to indicate the sorting
5. In a mobile view it says that the table is sorted by volume. Need to change to Date
mobile sorting
6. It would be auto-update the table (once per 30 sec?) when it is opened to show recent batces (not using manual page refresh)
autoupdate

7. Menu fonts looks like broken 
fonts
8. It would be great to add a copy icon closer to the TX hash. See how it is implemented for order/TX details pages: 
copy
9. 'View on Etherscan' in the menu sounds incorrect for Gnosis chain. It should be 'View on Gnosisscan'
10. Search is not working in the table
search is not working - disable
11. It would be great to add an anchor to the link and do not navigate a user to the Top tokens tab when refresh the page
image
12. Token icons with amount are broken on order details/user details/TX details pages in this PR: 
TX details
13. No USD estimation for Gnosis chain for a batch value
image
14. I think it would be great to show a full date on hover/click on the Date column. Something like 
date like
15. I'd add a bit more data in a mobile view. E.g. to show a token name close to a  token symbol. Also. menu looks weird as the last column there. 
menu, add symbols

**Comment by 0xaaiden** [comment]
Thanks @fleupold and @elena-zh for the detailed and extensive feedback (keep it coming!)
I've made all the styling changes mentioned above and fixed the search bar (filter applied on Tx. hash and solver ids).
@fleupold - for computing the batch value - it's now showing "Not Available" if the USD prices are not available for both the sell token or buy token. 

@elena-zh - I've removed the menu in mobile display as it was cluttering and didn't want to surcharge the screen. I've also added hot refetching every 30 seconds to update the table with the latest batches

I've attached a video below summarizing the changes.

https://github.com/cowprotocol/explorer/assets/30964272/fe72abe6-2cf5-44da-89a3-0b667cafeff9

File: .env.example [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
+# Enables mock mode (default = true)	
+MOCK=true	
 
-# Enables autoconnect for mock mode (default = true)
-AUTOCONNECT=true
+# Enables autoconnect for mock mode (default = true)	
+AUTOCONNECT=true	
 
-# Public IPFS gateway
-REACT_APP_IPFS_READ_URI=https://cloudflare-ipfs.com/ipfs
```
Comment (anxolin) [inline .env.example:8]: why the whitespaces?

File: src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    const timeOutId: NodeJS.Timeout = setTimeout(() => {
+      setIsFirstLoading(false)
+    }, timeOutMs)
+
+    return (): void => {
+      clearTimeout(timeOutId)
+    }
+  }, [batches, batches?.length])
```
Comment (anxolin) [inline src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx:32]: why do we need `batches?.length]` isn't batches enough?

File: src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  useEffect(() => {
+    let timeOutMs = 0
+    if (!batches) {
+      timeOutMs = DEFAULT_TIMEOUT
+    }
+
+    const timeOutId: NodeJS.Timeout = setTimeout(() => {
+      setIsFirstLoading(false)
+    }, timeOutMs)
```
Comment (anxolin) [inline src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx:27]: these logics seems overcomplicated, plus is strange that you are using this for the loading, but is only based on a timer... 

Wouldn't be cleaner design if the context would have a `loading: boolean` flag ? so responsibility is not in this component (the loading logic is not here anyways, so it doesn't feel is this component responsibility)

File: src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    return (): void => {
+      clearTimeout(timeOutId)
+    }
+  }, [batches, batches?.length])
+
+  return isFirstRender || isFirstLoading ? (
```
Comment (anxolin) [inline src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx:34]: do you handle the errors too?

File: src/apps/explorer/components/TokensTableWidget/index.tsx [hunk 5]
```diff
@@ -17,7 +20,12 @@ import CowLoading from 'components/common/CowLoading'
 import { EmptyItemWrapper } from 'components/common/StyledUserDetailsTable'
 import { ScrollBarStyle } from 'apps/explorer/styled'
 import { CardRow } from 'components/common/CardRow'
+import { useHistory, useLocation } from 'react-router-dom'
 
+const TABS = {
+  TOKENS: 1,
+  BATCHES: 2,
+}
```
Comment (anxolin) [inline src/apps/explorer/components/TokensTableWidget/index.tsx:28]: can we use a enum?

File: src/apps/explorer/components/TokensTableWidget/index.tsx [hunk 6]
```diff
@@ -71,39 +79,52 @@ const ExtraComponentNode: React.ReactNode = (
     <TablePagination context={TokensTableContext} fixedResultsPerPage />
   </WrapperExtraComponents>
 )
+interface Props {
+  networkId: BlockchainNetwork
+}
```
Comment (anxolin) [inline src/apps/explorer/components/TokensTableWidget/index.tsx:84]: nit: can you give a not so generic name for these props? 
you use it in `TokensTableWidget` so what about `TokensTableWidgetProps` and export it so other components can use it

Comment (anxolin) [reply src/apps/explorer/components/TokensTableWidget/index.tsx:84]: also, see next lines. Looks like a duplicate

File: src/apps/explorer/components/TokensTableWidget/index.tsx [hunk 7]
```diff
@@ -121,6 +142,14 @@ export const TokensTableWidget: React.FC<Props> = () => {
     setPageOffset(0)
   }, [networkId, setPageOffset])
 
+  useEffect(() => {
+    const handleHashChange = (): void => {
+      setSelectedTab(location.hash === '#batches' ? TABS.BATCHES : TABS.TOKENS)
+    }
+    window.addEventListener('hashchange', handleHashChange)
```
Comment (anxolin) [inline src/apps/explorer/components/TokensTableWidget/index.tsx:149]: mmmm, is this the reactRouter way of doing it?

File: src/apps/explorer/components/TokensTableWidget/index.tsx [hunk 8]
```diff
@@ -121,6 +142,14 @@ export const TokensTableWidget: React.FC<Props> = () => {
     setPageOffset(0)
   }, [networkId, setPageOffset])
 
+  useEffect(() => {
+    const handleHashChange = (): void => {
+      setSelectedTab(location.hash === '#batches' ? TABS.BATCHES : TABS.TOKENS)
```
Comment (anxolin) [inline src/apps/explorer/components/TokensTableWidget/index.tsx:147]: if you have an enum for this, you can use it also for checking the condition, so no need to add the string `'#batches'`

File: src/apps/explorer/components/TokensTableWidget/index.tsx [hunk 9]
```diff
@@ -143,12 +176,19 @@ export const TokensTableWidget: React.FC<Props> = () => {
       </EmptyItemWrapper>
     )
   }
-
+  const ExtraComponentNodeSearchBar: React.ReactNode = (
+    <div style={{ marginRight: '1rem' }}>
```
Comment (anxolin) [inline src/apps/explorer/components/TokensTableWidget/index.tsx:180]: pls, don't hardcode CSS

File: src/apps/explorer/components/TokensTableWidget/index.tsx [hunk 10]
```diff
@@ -143,12 +176,19 @@ export const TokensTableWidget: React.FC<Props> = () => {
...
+        <TableSearch query={query} setQuery={setQuery} />
+      </WrapperExtraComponents>
+    </div>
+  )
   return (
     <TableWrapper>
       <TokensTableContext.Provider
         value={{
           data: filterData(),
+          batchesData: filterBatches(),
```
Comment (anxolin) [inline src/apps/explorer/components/TokensTableWidget/index.tsx:191]: do we want to execute these function in each re-render?

File: src/apps/explorer/components/TokensTableWidget/index.tsx [hunk 11]
```diff
@@ -160,7 +200,17 @@ export const TokensTableWidget: React.FC<Props> = () => {
         }}
       >
         <ConnectionStatus />
-        <ExplorerCustomTab extraPosition={'bottom'} tabItems={tabItems(query, setQuery)} extra={ExtraComponentNode} />
+        <ExplorerCustomTab
+          extraPosition={'bottom'}
+          tabItems={tabItems()}
+          updateSelectedTab={(tabId: number): void => {
+            setSelectedTab(tabId)
+            history.push(tabId === TABS.BATCHES ? '#batches' : '#tokens')
```
Comment (anxolin) [inline src/apps/explorer/components/TokensTableWidget/index.tsx:208]: here you should use the enum... also can you do it in a more reactRouter way?

File: src/hooks/useGetBatches.ts [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+    const name =
+      trade.sellToken.address.toLowerCase() === '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'
+        ? 'Ethereum'
+        : trade.sellToken.name
+
+    return { name, symbol, address: trade.sellToken.address }
+  })
+  // Remove duplicates
+  return tokens.filter((token, index) => tokens.findIndex((t) => t.address === token.address) === index)
+}
```
Comment (anxolin) [inline src/hooks/useGetBatches.ts:244]: i didn't review fully this file. Needs to be cleaned up. Too big and hard to review.

Comment (anxolin) [reply src/hooks/useGetBatches.ts:244]: Also, i believe thegraph types should be auto-generated. 

Actually, this should already be part of the SDK (the autogeneration of types)

**Review by anxolin — COMMENTED** [review]
Thanks for your contribution!!

**Comment by anxolin** [comment]
One more thing, I actually wouldn't mind to make either the last batches the entry point, or to show both somehow (change the design slightly)

Easier is the first option. Just pinging @fairlighteth who might have an oppinion

**Comment by elena-zh** [comment]
Hey @0xaaiden , great job!
Still I have some nitpicks: 
1. Mobile view, 2 tokens in the batch: it would be gear to add a space between tokens
space
2. CoW in the Explorer should be with 'W'
cow
3. Goerli: no 'View on Etherscan' link that navigates to 'goerli.etherscan.tx_hash'
goerli - TX on Etherscan or hide
4. When change a network, the app stays on the Recent batches tab, but the anchor is removed from the link. I think, we should keep it :) 
not remove anchor
5. It would be great to indicate in the search field available search options 9n the placeholder. Currently, it shows a value for Tokens table
6. I have not been able to search by a solver address (however, IDK its full hash, just tried to search by this, and was not successful
image
7. I'd hide Copy action from the menu as currently it does nothing but navigating a user to the top on the Home page. Besides, not we have an icon neat a TX hash. 

Thanks!

File: .env.example (hunk 1 above)
Comment (0xaaiden) [reply .env.example:8]: Not sure why it happened!$, fixed!

File: src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx (hunk 3 above)
Comment (0xaaiden) [reply src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx:27]: That's the same logic used to load the Tokens too

File: src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx (hunk 4 above)
Comment (0xaaiden) [reply src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx:34]: It's handled in the batchesTable component,
```ts
if (!items || items.length === 0) {
      tableContent = (
        <tr className="row-empty">
          <td className="row-td-empty">
            <EmptyItemWrapper>
              No results found. <br /> Please try another search.
            </EmptyItemWrapper>
          </td>
        </tr>
```
 If there is not data returned, it shows "No results found."

**Comment by shoom3301** [comment]
Hi!
We don't see any activity in the PR for a long time.
Unfortunately, we have to close the PR since we are moving to another repo: https://github.com/cowprotocol/cowswap/pull/3579.
We will be happy if you continue the work here.
Thank you!

**Pull request closed without merging**